### ISSUE TITLE
echocon sample does not work

### ISSUE THREAD

**Issue by amoldeshpande** [microsoft/terminal#1824]
# Environment

```none
Windows build number: 18362.207
Windows Terminal version (if applicable):

Any other software? VS2019 with Windows 10 SDK 17763
```

# Steps to reproduce

Launch the sample

# Expected behavior

it should work

# Actual behavior

The call to SetConsoleMode fails with ERROR_INVALID_PARAMETER. If I ignore this error, the sample works as written. No idea if it would still work with an actual VT100 client. 

ETA: The ReadFile in PipeListener thread never returns while the child process is happily printing to the console window it was spawned with. 

So, I don't really understand how this thing illustrates anything about ConPTY.

**Comment by amoldeshpande** [comment]
hi @DHowett-MSFT, I see you tagged it with "help-wanted", and I would be happy to fix the sample if there was a way. 
As far as I can tell, the sample is using the relevant APIs as documented in MSDN and used in various Microsoft-authored articles and code snippets since ConPTY was announced.  

However,  things simply do not work (the SetConsoleMode call is a distraction in the sample and can be removed with no consequences to anything it demonstrates).  There is a conhost.exe spawned with the child process, but the output is not redirected to the pipes.

So, I am really at a loss in understanding the correct way of using the ConPTY API.  I tried downloading the Terminal source but gave up on using it to troubleshoot EchoCon,  when I realized Terminal won't run out of Visual Studio but needs to be packaged and has other hoops.

Also, looking at the Terminal source where it creates ConPTYs , it is identical to the EchoCon sample, so I'm back to square one in that respect.

TL:DR; I think there is more here than just a sample issue.

**Comment by DHowett-MSFT** [comment]
Hey @amoldeshpande,
Thanks for the comprehensive report! I'm actually not able to reproduce this.

I've modified the code to replace the escape character with `#` _only when it receives it on the output pipe_

```patch
diff --git a/samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp b/samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp
index 428f1e3e..2cc12131 100644
--- a/samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp
+++ b/samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp
@@ -183,6 +183,8 @@ void __cdecl PipeListener(LPVOID pipe)
         // Write received text to the Console
         // Note: Write to the Console using WriteFile(hConsole...), not printf()/puts() to
         // prevent partially-read VT sequences from corrupting output
+       for(int i = 0; i < dwBytesRead; ++i)
+               if(szBuffer[i] == '\x1b') szBuffer[i] = '#';
         WriteFile(hConsole, szBuffer, dwBytesRead, &dwBytesWritten, NULL);

     } while (fRead && dwBytesRead >= 0);
```

This is what I'm getting (which illustrates that the output from `ping` is being received on the output pipe):

```
#[25l#[2J#[m#[H#]0;C:\WINDOWS\SYSTEM32\ping.exe#[?25h
Pinging DHOWETT-DEV [::1] with 32 bytes of data:
Reply#[70X from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
Reply from ::1: time<1ms
#[25X
Ping statistics for ::1:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:#[10X
    Minimu#[46Xm = 0ms, Maximum = 0ms, Average = 0ms
```

Using `ENABLE_VIRTUAL_TERMINAL_PROCESSING` and passing the output pipe's contents _directly to standard out_ makes the sample look exactly like running ping.exe in conhost, you're right.

**Comment by amoldeshpande** [comment]
hmm, that really makes no sense to me. Are you running 18362 or some internal build ?  I also updated my sdk to 18362 (though that should not matter)

I get error 87 from SetConsoleMode failing and never hit the breakpoint I put in your for loop above.

I see  separate windows for ping.exe and echocon.exe (and in the debugger I can see conhost.exe launched as well)

what else can I do to troubleshoot ?

**Comment by DHowett-MSFT** [comment]
Hold up, do you have the "Use Legacy Console" option turned on here? I think `ENABLE_VIRTUAL_TERMINAL_PROCESSING` won't work in the legacy console, and that's why you're seeing ping spawn its own console window.

![image](https://user-images.githubusercontent.com/14316954/60933407-f44c6a00-a276-11e9-8362-120dae731581.png)

**Comment by amoldeshpande** [comment]
Yes ! that was it ! I didn't turn it on, so it must default to that mode.

I feel appropriately silly now.  Though in my defense, the blog articles with code snippets do not mention this or have calls to SetConsoleMode :-)

**Comment by DHowett-MSFT** [comment]
Nah, you're totally right that this is poorly documented. We actually just discovered it ourselves! #1838 :smile:
